### DISCUSSION TITLE
Question about setting the specific initial condition in the MOOSE simulation

### DISCUSSION THREAD

**Discussion by abc-hy** [idaholab/moose#23393] [Q&A Modules: Phase field] [answered]
Hi everyone,

I have a question about setting the initial condition in MOOSE simulation.
I want to set the initial condition as the following shape, which is inside the red dotted line. How can I get its concentration from the files and put it in the new simulation? Can anyone give me some advice?
![image (6)](https://user-images.githubusercontent.com/63679626/217939730-ac7d1c9a-d65a-48cc-9e8f-84f174683e5e.png)

Thank you very much.
Best,

**Comment by GiudGiud** [comment]
Hello

so you have this profile in a file? What kind of file? Exodus or CSV we can use SolutionUserObject or PropertyReadFile respectively

Guillaume

**Reply by abc-hy** [reply]
it is Exodus. But I also need to cut the model, only use the area inside the red-dotted line, how to realize it?

**Reply by GiudGiud** [reply]
You can use a ParsedSubdomainGenerator to set subdomains along straight lines like this
Then you can block restrict the setting of the initial conditions

**Reply by abc-hy** [reply]
So, do I need to change the [Mesh] block? I am a little bit confused....
I attached my files here, could you have a look at it? How to load the data?
Also, I converted the Exodus file to cvs file, and I also attached here.
[df_0d6_t_1_135.csv](https://github.com/idaholab/moose/files/10703525/df_0d6_t_1_135.csv)
[df0d6_t_1_R_txt.txt](https://github.com/idaholab/moose/files/10703532/df0d6_t_1_R_txt.txt)

**Reply by GiudGiud** [reply]
You can add another generator in the Mesh block to create that new subdomain. 

If you have an exodus file with the data please have a look at the SolutionUserObject. 

If the element ids aren’t exactly the same in your file and the new simulation this might not work btw. Please let us know once you’ve set it up

**Reply by abc-hy** [reply]
I did not load the shape successfully...
Since what I want to do is to add a perturbation to my simulation at the beginning, can I instead use this 'noise' function as the perturbation in my simulation?
![image](https://user-images.githubusercontent.com/63679626/218215739-b7d0f1c9-dd2a-4574-842b-7c49f7c94f82.png)

**Reply by abc-hy** [reply]
Hi,

Do you know what is source_nodal?
![image](https://user-images.githubusercontent.com/63679626/218788766-503e7c2f-166a-45da-8df6-0a2adf9b6feb.png)
Also, If I have 2 variables, how should I use 'SolutionUserObject' to read them? I only find how to read one variable from the examples.

Thank you.
Best,

**Reply by GiudGiud** [reply]
Hello

Which input file is this from?
Also you should paste text instead of screenshots. 

Guillaume

**Reply by abc-hy** [reply]
From SolutionUserObject website
https://mooseframework.inl.gov/source/userobject/SolutionUserObject.html

**Reply by GiudGiud** [reply]
this is the name of the variable in the solution file. If you navigate to the test/tests/auxkernels/solution_aux folder and open the exodus file you ll see that variable.

**Reply by abc-hy** [reply]
Ok, so the source_nodal is only a name for one variable? How can I read two variables?

**Reply by GiudGiud** [reply]
yes the parameter is a vector, you can put as many variables as there are in the files in the parameter with this syntax
'var1 var2'

**Reply by abc-hy** [reply]
I used this 'c w', but I get error messages. It seems this does not work.
[UserObjects]
  [./soln]
    type = SolutionUserObject
    mesh = df0d6_t_1_out.e-s136
    system_variables = 'c w'
    execute_on = INITIAL
    scale = '1 1 0'

  [../]
[]

[Functions]
  [./solution_fcn]
    type = SolutionFunction
    from_variable = 'c w'
    solution = soln
  [../]
[]
I got this error message.
![image](https://user-images.githubusercontent.com/63679626/218803304-a8a6f2a4-1f25-4129-8e28-4b440a85d485.png)

**Reply by GiudGiud** [reply]
This is because the solution function has from_variable (no s, it s singular) with 'c w' instead of a single variable 'c'. 
This makes sense, a function cannot reproduce the shape of two variables. It can only do one

Please examine the parameters before setting them. They are listed at the bottom of every object's documentation
For example here, on this page:
https://mooseframework.inl.gov/source/functions/SolutionFunction.html

**Reply by abc-hy** [reply]
You mean I don't need to use solution function?

**Reply by GiudGiud** [reply]
no you need to fix the parameters for it.

**Reply by abc-hy** [reply]
Are the parameters not c and w? They are the variables in my files.

**Reply by GiudGiud** [reply]
Just have two solutionFunction, one with c and one with w

**Reply by abc-hy** [reply]
Do I need to add other AuxKernel? Since the two variables are the same in the previous simulations

**Reply by GiudGiud** [reply, marked as answer]
You should probably just be using two of these
https://mooseframework.inl.gov/source/auxkernels/SolutionAux.html

**Reply by abc-hy** [reply]
![image](https://user-images.githubusercontent.com/63679626/218817852-bc75ca60-d1e6-4f0e-b5ba-e5df56892ac0.png)
I got this error message by using two Solution function, shown as below:
![image](https://user-images.githubusercontent.com/63679626/218817980-6057b901-5d45-4df1-b2b1-c25de27a5c6b.png)
